ssl/ssl_conf.c

```diff
@@ -599,13 +599,15 @@ static int cmd_DHParameters(SSL_CONF_CTX *cctx, const char *value)
                                             sslctx->libctx, sslctx->propq);
         if (decoderctx == NULL)
             goto end;
-        while (!OSSL_DECODER_from_bio(decoderctx, in) && !BIO_eof(in)) {
-            ERR_clear_error();
-        }
+        ERR_set_mark();
+        while (!OSSL_DECODER_from_bio(decoderctx, in) && !BIO_eof(in));
```

Comment:
You need to put an additional `if (dhpkey != NULL) break;` inside the `while` loop or just add `dhpkey == NULL` as an additional condition of the loop. There is no point in further reading from the BIO if you encounter parameters. I do not even know if this would work correctly if you had something else in the file after the parameters.